make WeakKeyDict finalizers thread/async-safe

vtjnash opened the pull request:
~~The finalizers for `AbstractRemoteRef` were corrupting the identity of those objects. By not doing that, I think this may provide a fix for some of the recent parallel bugs. Along the way, I implemented making Dict finalizer-safe (while I thought that might be related to the currrent bug).
old commits: 9b5fa5d, feaf59f, 8d1970e~~

this makes WeakKeyDict finalizer usage gc-safe, which should fix numerous testing issues from `client_ref` access race-conditions

vtjnash committed:
improve efficiency of Dict and accuracy of age (nee dirty) flag

ref comment in #15923

yuyichao commented on src/gc.c:
```diff
@@ -539,16 +540,28 @@ static void run_finalizers(void)
+    else if (prev && !on) {
+        // enable -> disable
+        jl_atomic_fetch_add(&jl_finalizers_disabled_counter, 1);
+        // TODO: check if finalizers are running and wait for them to finish
```
Is this really necessary? Since finalizer runs can nest and can run on multiple threads, I feel like waiting for all finalizers to finish is asking for dead lock. (e.g. if this is called by a finalizer on two different threads because both of them modifies a dict).

For avoiding corruption with multithreading it is still necessary to use explicit locks. It is only the case if the finalizer runs on the same thread that you can't really protect with a lock and requires this API to disable the finalizer on the same thread.

yuyichao replied:
In another word, I think we probably only need a thread local `disable_finalizers` and the caller of this function is expected to do the necessary thread synchronization.

vtjnash replied:
I think we may need to think about ways of localizing finalizers to the appropriate thread. technically though, this lock condition might only need to control access to the finalizer list, but would not need to be held while running a finalizer.

if we stop running finalizers on the same thread, then I guess the more standard implementation would be to swap this for explicit internal Dict locks. but currently, I think we would need both (to avoid having the finalizer try to grab the non-recursive lock we are already holding and deadlock).

yuyichao replied:
Whether we need a Dict lock or not depends on what's the thread safety guarantee we provide for stdlib. Currently, the user always has to synchronize access to a dict from multiple threads and there's nothing special about whether it is running from within a finalizer or not and the only feature Base has to provide is that finalizer running on the same thread will not corrupt the data structure.

If, however, we begin to guarantee thread safety of Base data structures it would indeed be better to use that for finalizer synchronization too.

vtjnash replied:
true. but I think when we do that I think we either need to make finalize-disable part of acquiring a lock, or run all finalizers on a separate thread. for now, I've been tackling those concerns independently. I can understand the argument though that multi.jl might be responsible for synchronizing the Dict access to client_refs. Somehow WeakKeyDict would need to be compliant and use the same lock.

vtjnash commented on src/julia.h:
```diff
@@ -1492,6 +1495,8 @@ STATIC_INLINE void jl_eh_restore_state(jl_handler_t *eh)
 #endif
     JL_SIGATOMIC_END();
+    jl_finalizers_enable(eh->finalizers_enabled);
+    jl_gc_enable(eh->gc_enabled);
```
this changes the behavior of the test in `misc` for gc being disabled after `try; gc_enable(false); end`. I'm actually inclined to think the the new behavior is better (and fix the test / document that `gc_enable` is scoped to the nearest try/catch blocks), but I'm open to counter-arguments.

yuyichao replied:
I think the new one is better in general. `gc_enable` really shouldn't be used in julia code so it's hard to say what's the expected usage pattern....

FYI, this adds another conditional safepoint to the beginning of the `catch` block =)

vtjnash replied:
I think gc is OK, I mainly didn't want finalizers or other non-sigatom regions (because they can corrupt `jl_exception_in_transit` + the `catch_backtrace()`, or and possible also skip `finally` blocks). i'm starting to think we need to make a dedicated thread for running all finalizers.

tkelman commented:
is this branch/pr still necessary after you pushed part of it to master? if not, please close the pr and delete the branch

vtjnash committed:
split weakkeydict into a separate file

vtjnash committed:
move Thread support much earlier in bootstrapping

vtjnash committed:
make WeakKeyDict finalizer usage gc-safe

also use this `client_refs.lock` to protect other data-structures from
being interrupted by finalizers, in the multi.jl logic

we may want to start indicating which mutable data-structures are safe
to call from finalizers, since generally that isn't possible

to make a finalizer API gc-safe, that code should observe the standard
thread-safe restrictions (there's no guarantee of which thread it'll run on),

plus, if the data-structures uses locks for synchronization,
use the `islocked` pattern (demonstrated herein) in the `finalizer`
to re-schedule the finalizer when the mutable data-structure is not
available for mutation.
this ensures that the lock cannot be acquired recursively,
and furthermore, this pattern will continue to work if finalizers
get moved to their own separate thread.

close #14445
fix #16550
reverts workaround #14456 (shouldn't break #14295, due to new locks)
should fix #16091 (with #17619)

vtjnash commented:
PR re-written to synchronize only WeakKeyDict, and re-schedule finalizers rather than disabling them

(updated PR based on #17590)

KristofferC commented on doc/stdlib/parallel.rst:
```diff
@@ -800,6 +782,105 @@ will) change in the future.
+
+   Test-and-test-and-set spin locks are quickest up to about 30ish contending threads. If you have more contention than that, perhaps a lock is the wrong way to synchronize.
+
+   See also SpinLock for a version that permits recursion.
```
RecursiveSpinLock

vtjnash replied:
This is auto-generated, I just forgot to rerun the script.

KristofferC replied:
I know it is but I cba to find the actual source file :)

tkelman commented on base/multi.jl:
```diff
@@ -610,34 +621,26 @@ type RemoteChannel{T<:AbstractChannel} <: AbstractRemoteRef
-            # is fixed.
-            client_refs[r] = false
-        end
+    if r.where > 0 # Handle the case of the finalizer having being called manually
```
having been called

tkelman commented on base/multi.jl:
```diff
@@ -610,34 +621,26 @@ type RemoteChannel{T<:AbstractChannel} <: AbstractRemoteRef
-            client_refs[r] = false
-        end
+    if r.where > 0 # Handle the case of the finalizer having being called manually
+        islocked(client_refs) && return finalizer(r, finalize_ref) # delay finalizer for later when its not already locked
```
it's not already locked

tkelman commented on base/lock.jl:
```diff
@@ -95,6 +95,25 @@ function unlock(rl::ReentrantLock)
+    end
+end
+
+function trylock(f, l)
```
these forms could use unit tests

vtjnash replied:
added

tkelman commented on base/weakkeydict.jl:
```diff
@@ -2,62 +2,74 @@
     k = convert(K, key)
-    if is(wkh.deleter, identity)
-        wkh.deleter = x->weak_key_delete!(t, x)
+    finalizer(k, wkh.finalizer)
```
so this may or may not add the finalizer to the input depending ob whether the convert aliased?

tkelman commented on base/locks.jl:
```diff
@@ -1,33 +1,55 @@
 # contending threads. If you have more contention than that, perhaps
 # a lock is the wrong way to synchronize.
+"""
+    SpinLock()
```
if this is how this is expected to be called, why isn't it typealias'ed the other way around?

vtjnash replied:
Added by @kpamnany in f3402da30706b2e55e8751b5a8460b001db88b74

tkelman replied:
sure but you're adding the docstring now, and chose to use the typealias name in the docstring instead of the actual type name?

kpamnany replied:
I don't understand this comment? I want to potentially have multiple implementations of ``SpinLock``, selecting a particular one depending on platform. But I want user code to use just ``SpinLock``.

tkelman replied:
then the docstring should be for spinlock as written, and actually attached to the spinlock name instead of the tataslock name

tkelman commented on base/weakkeydict.jl:
```diff
@@ -0,0 +1,81 @@
+
+function setindex!{K}(wkh::WeakKeyDict{K}, v, key)
+    k = convert(K, key)
+    finalizer(k, wkh.finalizer)
```
collapsed but still applicable:

> so this may or may not add the finalizer to the input depending on whether the convert aliased?

vtjnash replied:
yes, that's the point of it being a WeakKeyDict

tkelman commented on base/weakkeydict.jl:
```diff
@@ -0,0 +1,81 @@
+    return (kv, (i, gc_token))
+end
+
+function filter!(f, d::Union{ObjectIdDict,Dict,WeakKeyDict})
```
union not necessary here, given `function filter!(f, d::Union{ObjectIdDict,Dict})` is still defined in `dict.jl` ?

tkelman commented on base/weakkeydict.jl:
```diff
@@ -0,0 +1,81 @@
+    end
+end
+
+get{K}(wkh::WeakKeyDict{K}, key, default) = lock(wkh) do; return get(wkh.ht, key, default); end
```
I think these would look better as `lock(() -> get(wkh.ht, key, default), wkh)` rather than 1-liner do blocks
